### ISSUE TITLE
Make assert more verbose

### ISSUE THREAD

**Issue by privat** [nitlang/nit#1395]
The engine could try to give more information when a assert fail.

Currently

``` nit
var a = "hello"
assert a == "world"
# Runtime error: Assert failed
```

That is not very informative, the dev has then to debug and print some information; usually, with an else because it is easy to and and intrusive

``` nit
var a = "hello"
assert a == "world" else print "a=`{a}`; want `hello`"
# a=`hello`; want `hello`
# Runtime error: Assert failed
```

Maybe, a way could be to provide `asserts` statement with a default `else` that inspect and print more useful things.

**Comment by Morriar** [comment]
The @privat could try to give more information about the information to provide when an assert fails.

**Comment by privat** [comment]
updated the description for the @Morriar

**Comment by Morriar** [comment]
I'm not sure about the default else because most of the `else` in `assert` are used to display more information. You don't want the programmer to chose between a clear message and the expected value.

I will go a systematic print of the expected and received value even if a `else` is there.

**Comment by R4PaSs** [comment]
Can I revive this ? More and more do I see newcomers surprised when an assert fails due to a lack of information being shown.

The `assert else` method can be a solution, although I suspect most programmers are as lazy as I am (and this solution forces us to update every assert in `core` to support such a message, I'm not ready for this), therefore we might not have the will to do so every time we write an `assert`.

I find the solution proposed by @Morriar to be quite elegant, maybe an error message with an emphasis on the ~~god~~ *statement* that failed.

Such a program:

~~~nit
var s = "String"

for i in [0 .. s.length] do print s[i]
~~~

Which is currently producing this message:

~~~sh
S
t
r
i
n
g
Runtime error: Assert failed (lib/core/text/flat.nit:664)
-------------------------------------------------
--   Stack Trace   ------------------------------
-------------------------------------------------
` fatal_exit
` core::flat$ASCIIFlatString$Text::[] (lib/core/text/flat.nit:663)
` assert_fail::assert_fail$core::Sys$main (assert_fail.nit:1)
` main
` __libc_start_main
` _start
` 
-------------------------------------------------
~~~

Could instead produce a message like this one:

~~~sh
S
t
r
i
n
g
Runtime error: Assert failed (lib/core/text/flat.nit:664)

assert idx < _bytelen and idx >= 0
       ~~~~~~~~~~~~~~
-------------------------------------------------
--   Stack Trace   ------------------------------
-------------------------------------------------
` fatal_exit
` core::flat$ASCIIFlatString$Text::[] (lib/core/text/flat.nit:663)
` assert_fail::assert_fail$core::Sys$main (assert_fail.nit:1)
` main
` __libc_start_main
` _start
` 
-------------------------------------------------
~~~

Thus giving a better clue on what happened to the end-user.

**Comment by privat** [comment]
This is something I'm currently working on. But I need a way to do things correctly.
The main issue is to have enough expressiveness to manipulate the code to refer to itself.
for instance, in your example you break the condition apart and isolate the left operand of the `and`.

another thing that worries me is the size inflation of the executable if I have to embed parts of original code source in the final binary.

**Comment by privat** [comment]
closed by #2541
